src/il_representations/algos/__init__.py

```diff
@@ -272,25 +275,32 @@ def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
 
-class Autoencoder(RepresentationLearner):
+class Jigsaw(RepresentationLearner):
     """
-    A basic autoencoder that tries to reconstruct the
-    current frame, and calculates an MSE loss over current frame pixels,
-    using reconstruction loss.
+    A basic Jigsaw task that requires a network to solve Jigsaw puzzles.
     """
     def __init__(self, **kwargs):
         encoder_kwargs = kwargs.get('encoder_kwargs') or {}
         encoder_cls_key = encoder_kwargs.get('obs_encoder_cls', None)
-
-        algo_hardcoded_kwargs = dict(encoder=VAEEncoder,
-                                     decoder=PixelDecoder,
+        _this_file_dir = os.path.dirname(os.path.abspath(__file__))
+        data_root = os.path.abspath(os.path.join(_this_file_dir, '../../../'))
+        permutation_filename = 'jigsaw_permutations_1000.npy'
+        permutation_file = \
+            glob.glob(f'{data_root}/**/{permutation_filename}', recursive=True)[0]
```

Comment:
Is there a reason to have these stored in a file? If the aim is just to obtain determinism, then we could use something like this:

```python
import numpy as np

permutations_seed = 42  # change to get different set of permutations
n_permutations = 1000
n_tiles = 9

assert n_permutations < np.math.factorial(n_tiles), \
    f"cannot get {n_permutations} permutations with only {n_tiles} tiles"
rng = np.random.RandomState(permutations_seed)
unique_permutations = set()
while len(unique_permutations) < n_permutations:
    # We iteratively add permutations to a set in order to avoid duplicates.
    # With n_tiles=9, n_permutations=1000 we are highly likely to
    # get at least one duplicate
    new_permutation = tuple(rng.permutation(n_tiles))
    # uncomment this to see duplicates
    # if new_permutation in unique_permutations: print('dupe!')
    unique_permutations.add(new_permutation)
permutations = np.asarray(sorted(unique_permutations))
# shuffle the permutations so there's less structure in class IDs
permutations = permutations[rng.permutation(len(permutations))]
```

Reply:
Since the original Jigsaw paper requires that the 1000 permutations be sufficiently different (i.e., have a large hamming distance among each other), generating these permutations requires calculating the hamming distance once per iteration, which can take a long time as the size of the list of selected permutations grows (e.g., it takes 40 minutes on perceptron). So I just saved this list as a file and reload it every time. I have added a function into `utils.py` if people want to generate it again.

```
    all_perms = np.array(list(itertools.permutations(list(range(n_tiles)), n_tiles)))

    perms = []
    for i in range(n_perms):
        if i == 0:
            j = np.random.randint(len(all_perms))
            perms = np.array(all_perms[j]).reshape([1,-1])
        else:
            perms = np.concatenate([perms, all_perms[j].reshape([1,-1])], axis=0)

        all_perms = np.delete(all_perms, j, axis=0)
        
        # Calculate selected permutations' distance with all possible
        # permutations, and select the one that are most different from
        # selected ones. This is to make sure the generated jigsaw puzzles are
        # sufficiently different from each other.
        # This takes a long time when len(perms) grows.
        D = cdist(perms, all_perms, metric='hamming').mean(axis=0).flatten()
        j = D.argmax()
```

Reply:
Okay, got it. Thanks!